tests/integration/test_tenants.py

```diff
@@ -1,89 +1,67 @@
-        "dateTimeStart": Time.yesterday_iso(),
-        "unitNum": "413",
-    }
-
-    response = client.post(endpoint, json=newTenant, headers=valid_header)
-
-    assert is_valid(response, 201)  # CREATED
-    assert response.json["firstName"] == "Jake"
-
-    response = client.post(endpoint, json=newTenantWithLease, headers=valid_header)
-    assert is_valid(response, 201)
-    assert response.json["unitNum"] == "413"
-
-    response = client.post(endpoint, json=newTenant, headers=valid_header)
 
 
 @pytest.mark.usefixtures("client_class", "empty_test_db")
-class TestTenant:
+class TestGet:
     def setup(self):
         self.endpoint = "/api/tenants"
 
-    def test_update_tenant(self, empty_test_db, valid_header, create_tenant):
+    def test_get_tenant(self, valid_header, create_tenant):
         tenant = create_tenant()
-        updated_fields = {"archived": True}
 
-        with patch.object(TenantModel, "update", return_value=tenant) as mock_update:
-            response = self.client.put(
-                f"{self.endpoint}/{tenant.id}",
-                json=updated_fields,
-                headers=valid_header,
-            )
+        response = self.client.get(f"{self.endpoint}/{tenant.id}", headers=valid_header)
+
+        assert response.json == tenant.json()
+
+
+@pytest.mark.usefixtures("client_class", "empty_test_db")
+class TestPut:
+    def setup(self):
+        self.endpoint = "/api/tenants"
+
+    def test_update_tenant(self, valid_header, create_tenant):
+        tenant = create_tenant()
+        updated_fields = {"archived": True}
 
-        mock_update.assert_called_once_with(
-            schema=TenantSchema, id=tenant.id, payload=updated_fields
+        response = self.client.put(
+            f"{self.endpoint}/{tenant.id}",
+            json=updated_fields,
+            headers=valid_header,
         )
-        assert response.status_code == 200
+
         assert response.json == tenant.json()
+
+
+@pytest.mark.usefixtures("client_class", "empty_test_db")
+class TestGetAll:
+    def setup(self):
+        self.endpoint = "/api/tenants"
+
+    def test_get_all_tenants(self, valid_header, create_tenant):
+        create_tenant()
+        create_tenant()
+        all_tenants = TenantModel.query.json()
+
+        response = self.client.get(self.endpoint, headers=valid_header)
+
+        assert response.json["tenants"] == all_tenants
+
+
+@pytest.mark.usefixtures("client_class", "empty_test_db")
+class TestCreate:
+    def setup(self):
+        self.endpoint = "/api/tenants"
+        self.new_tenant = {
+            "firstName": "Testy",
+            "lastName": "McTesterson",
+            "phone": "1234567890",
+            "archived": False,
+        }
+
+    def test_create_tenant(self, valid_header):
+        response = self.client.post(
+            self.endpoint, json=self.new_tenant, headers=valid_header
+        )
+
+        assert response.json == TenantModel.find(response.json["id"]).json()
```

Comment:
Just like with the `get_all_tenants` test, the logic here is identical to the logic in the endpoint.

For this one, we should probably be patching the `create` method of the `TenantModel` and asserting that it was called once with the correct arguments, like the `test_update_tenant` test was doing.

If you need an example, you can take a look at the `test_create_lease` test:
https://github.com/codeforpdx/dwellinglybackend/blob/6028ed50aa66bf038b24fafcaf46d6dfbbef1072/tests/integration/test_leases.py#L42-L50

Let me or @NickSchimek know if you have any questions about how or why to do this!

Reply:
I'm getting the same 

```
TypeError: Object of type MagicMock is not JSON serializable
``` 
for this one, and it is modeled on the test_create_lease 

```python
        with patch.object(TenantModel, "create") as mock_create:
            response = self.client.post(
                self.endpoint, json=self.new_tenant, headers=valid_header
            )
```
I'm completely lost on this, but  I've been swamped with interviews/other projects the last couple of weeks, so my brain is basically tapioca at this point.

Reply:
Hm...now that I've taken a closer look at the Tenants endpoint, I think this create action is a lot harder to mock than I originally thought. Sorry for making such a difficult suggestion! 😅 

You can get around the error you are receiving by providing something that _is_ json serializable as the `return_value` of the mock, but the lease creation logic that's also in this `post` method messes it all up even when you do that.  Perhaps we can figure out a better way to test this on Wednesday or at a later date.  

Either way, however, so long as that lease creation logic is in there, we should also have a test checking that a lease gets created and associated with the tenant when `dateTimeEnd`, `dateTimeStart` and `propertyID` are present in the payload.  I believe the original `testTenantsPOST` test did this to some extent.

Reply:
No problem.  I played with it for a while, but I don't feel like I have solid grasp on how patch works with beyond the straight forward examples we've used.  I will try and do a bit more research before Wednesday's meeting in case we want to take these to the next level.

How do we feel about these tests?:

```python

class TestCreate:
    def setup(self):
        self.endpoint = "/api/tenants"
        self.new_tenant = {
            "firstName": "Testy",
            "lastName": "McTesterson",
            "phone": "1234567890",
            "archived": False,
        }

    def test_create_tenant(self, valid_header):
        response = self.client.post(
            self.endpoint, json=self.new_tenant, headers=valid_header
        )

        assert response.status_code == 201
        assert response.json["firstName"] == self.new_tenant["firstName"]
        assert response.json["lastName"] == self.new_tenant["lastName"]
        assert response.json["phone"] == self.new_tenant["phone"]

    def test_create_tenant_with_lease(self, valid_header, create_property, create_lease):
        lease = {
            'occupants': 47,
            'dateTimeStart':"2021-06-19T07:00:00.000Z",
            'dateTimeEnd':"2022-07-16T07:00:00.000Z",
            'unitNum': '5351',
            "propertyID": create_property().id
        }

        response = self.client.post(
            self.endpoint, json={**self.new_tenant, **lease}, headers=valid_header
        )

        assert response.status_code == 201
        assert response.json["firstName"] == self.new_tenant["firstName"]
        assert response.json["lastName"] == self.new_tenant["lastName"]
        assert response.json["phone"] == self.new_tenant["phone"]
        assert response.json["unitNum"] == lease["unitNum"]
        assert response.json["occupants"] == lease["occupants"]
```

Reply:
Also, I need to dig into it some more, but it appears to be a problem with the returned data from this endpoint.  The lease information is part of the main object, not a part of the "lease" object:

```python
{'id': 1, 'firstName': 'Testy', 'lastName': 'McTesterson', 'fullName': 'Testy McTesterson', 'phone': '1234567890', 'lease': '', 'staff': [], 'created_at': '06/15/2021 18:02:28', 'updated_at': '06/15/2021 18:02:28', 'archived': False, 'occupants': 47, 'propertyID': 1, 'unitNum': '5351'}
```

I believe it should look like: 
```python
{'id': 1, 'firstName': 'Testy', 'lastName': 'McTesterson', 'fullName': 'Testy McTesterson', 'phone': '1234567890', 'staff': [], 'created_at': '06/15/2021 18:02:28', 'updated_at': '06/15/2021 18:02:28', 'archived': False,'lease': {'occupants': 47, 'propertyID': 1, 'unitNum': '5351'}}

I will look more into it, but is this a known problem, or a problem at all?